src/net_processing.cpp

```diff
@@ -94,144 +94,156 @@ void static ProcessGetData(CNode *pfrom, const Consensus::Params &consensusParam
-                // it's available before trying to send.
-                if (fSend && (mi->nStatus & BLOCK_HAVE_DATA))
-                {
-                    // Send block from disk
-                    CBlock block;
-                    if (!ReadBlockFromDisk(block, mi, consensusParams))
+                        // disconnect node
+                        pfrom->fDisconnect = true;
+                        fSend = false;
+                    }
+                    // Avoid leaking prune-height by never sending blocks below the
+                    // NODE_NETWORK_LIMITED threshold.
+                    // Add two blocks buffer extension for possible races
+                    if (fSend && !pfrom->fWhitelisted &&
+                        ((((nLocalServices & NODE_NETWORK_LIMITED) == NODE_NETWORK_LIMITED) &&
+                            ((nLocalServices & NODE_NETWORK) != NODE_NETWORK) &&
+                            (chainActive.Tip()->nHeight - mi->nHeight > (int)NODE_NETWORK_LIMITED_MIN_BLOCKS + 2))))
                     {
-                        // its possible that I know about it but haven't stored it yet
-                        LOG(THIN, "unable to load block %s from disk\n",
-                            mi->phashBlock ? mi->phashBlock->ToString() : "");
-                        // no response
+                        LOG(NET, "Ignore block request below NODE_NETWORK_LIMITED threshold from peer=%d\n",
+                            pfrom->GetId());
+                        // disconnect node and prevent it from stalling (would
+                        // otherwise wait for the missing block)
+                        pfrom->fDisconnect = true;
+                        fSend = false;
                     }
-                    else
+                    // Pruned nodes may have deleted the block, so check whether
+                    // it's available before trying to send.
+                    bool fHaveData = false;
                     {
-                        if (inv.type == MSG_BLOCK)
-                        {
-                            pfrom->blocksSent += 1;
-                            pfrom->PushMessage(NetMsgType::BLOCK, block);
-                        }
-                        else if (inv.type == MSG_THINBLOCK && pfrom->xVersion.as_u64c(XVer::BU_XTHIN_VERSION) < 2 &&
-                                 pfrom->ThinBlockCapable())
-                        {
-                            // TODO: This code path enables backward compatibility for older BU nodes
-                            // and can be removed in the future.
-                            LOG(THIN, "Sending thinblock via getdata message\n");
-                            SendXThinBlock(MakeBlockRef(block), pfrom, inv);
-                        }
-                        else if (inv.type == MSG_CMPCT_BLOCK)
+                        READLOCK(cs_mapBlockIndex);
+                        fHaveData = (mi->nStatus & BLOCK_HAVE_DATA);
```

Comment:
If the system can tolerate fHaveData being incorrect then we don't need the read lock.  If the system can't tolerate it being incorrect then the lock needs to be extended to cover ReadBlockFromDisk.  Otherwise the record could be deleted and mi->nStatus & BLOCK_HAVE_DATA changed after the lock is released but before the record is read from disk.

For performance reasons, it makes sense to allow fHaveData to be incorrect and gracefully degrade functionality (in this case, don't provide the block) in that rare case.  So I don't think that we need the lock here.

Reply:
I think you're right, I think we can tolerate a degraded mode so we can remove this lock.